crates/iota-data-ingestion-core/src/reducer.rs

```diff
@@ -0,0 +1,154 @@
+    // store unprocessed progress messages from workers.
+    let mut unprocessed = HashMap::new();
+    // buffer to accumulate results before passing them to the reducer.
+    // The size of this batch is dynamically determined by the reducer's
+    // `should_close_batch` method.
+    let mut batch = vec![];
+    // track the latest processed checkpoint number for reporting progress
+    // after each chunk of messages is received from the stream.
+    let mut progress_update = None;
+
+    loop {
+        tokio::select! {
+            _ = &mut shutdown => break,
+            // get the available progress from workers as a chunks
+            Some(update_batch) = stream.next() => {
+                unprocessed.extend(update_batch.into_iter());
+                // Process messages sequentially based on checkpoint sequence number.
+                // This ensures in-order processing and maintains progress integrity.
+                while let Some(message) = unprocessed.remove(&current_checkpoint_number) {
+                    if let Some(ref reducer) = reducer {
+                        // reducer is configured, collect messages in batch based on
+                        // `reducer.should_close_batch` policy, once a batch is collected it gets
+                        // committed and a new batch is created with the current message.
+                        if reducer.should_close_batch(&batch, Some(&message)) {
+                            reducer
+                                .commit(std::mem::take(&mut batch))
+                                .await?;
+                            batch = vec![message];
+                            progress_update = Some(current_checkpoint_number);
+                        } else {
+                            // Add message to existing batch since no commit needed
+                            batch.push(message);
+                        }
+                    }
+                    current_checkpoint_number += 1;
+                }
+                // Handle final batch processing
+                match reducer {
+                    Some(ref reducer) => {
+                        // Check if the final batch should be committed.
+                        // None parameter indicates no more messages available
+                        if reducer.should_close_batch(&batch, None) {
+                            reducer
+                                .commit(std::mem::take(&mut batch))
+                                .await?;
+                            progress_update = Some(current_checkpoint_number);
+                        }
+                    }
+                    None => progress_update = Some(current_checkpoint_number),
+                }
+                // report progress update to executor
+                if let Some(watermark) = progress_update {
+                    executor_progress_sender
+                        .send(WorkerPoolStatus::Running((task_name.clone(), watermark)))
```

Comment:
Just to make sure, the interpretation of watermark here is that it is id of last checkpoint that is NOT yet processed, right?

Reply:
It represents the last successfully processed checkpoint number.

Reply:
@tomxey 

After inspecting the logic closer I can confirm that your initial assumption is indeed correct! 

The `watermark` does represent the latest checkpoint not yet processed, in this way upon restart `ProgressStore::load` function will fetch that watermark value which will be the `current_checkpoint_number` from which all workers will start fetching and processing checkpoints.

Reply:
Got it, added another comment as a follow up.